File: src/commands/migrate.ts
import { Command } from 'commander';
import { mkdir, rename, readdir, rmdir } from 'node:fs/promises';
import { join, dirname, resolve, sep } from 'node:path';
import { execute } from '../core/executor.js';
import {
  readMetaConfig,
  getMetaDir,
  fileExists,
  addToGitignore,
  removeFromGitignore,
} from '../core/config.js';
import { findGitRepos } from '../core/discover.js';
import * as output from '../core/output.js';

interface MigrateOptions {
  dryRun?: boolean;
}

interface Move {
  from: string;
  to: string;
}

interface Conflict {
  path: string;
  found: string | null;
}

async function pruneEmptyParents(metaDir: string, movedFrom: string): Promise<void> {
  const root = resolve(metaDir);
  let dir = resolve(dirname(join(metaDir, movedFrom)));

  while (dir !== root && dir.startsWith(root + sep)) {
    let entries: string[];
    try {
      entries = await readdir(dir);
    } catch {
      return;
    }
    if (entries.length > 0) {
      return;
    }
    await rmdir(dir);
    dir = dirname(dir);
  }
}

async function getRemoteUrl(dir: string): Promise<string | null> {
  const result = await execute('git remote get-url origin', { cwd: dir });
  if (result.exitCode === 0 && result.stdout) {
    return result.stdout.trim();
  }
  return null;
}

async function mapUrlsToCurrentPaths(
  metaDir: string,
  ignore: string[]
): Promise<{ urlToPath: Map<string, string>; ambiguousUrls: Set<string> }> {
  const repoPaths = await findGitRepos(metaDir, ignore);
  const urlToPath = new Map<string, string>();
  const ambiguousUrls = new Set<string>();

  for (const repoPath of repoPaths) {
    const url = await getRemoteUrl(join(metaDir, repoPath));
    if (!url) {
      continue;
    }
    if (urlToPath.has(url)) {
      ambiguousUrls.add(url);
      continue;
    }
    urlToPath.set(url, repoPath);
  }

  return { urlToPath, ambiguousUrls };
}

export async function migrateCommand(options: MigrateOptions = {}): Promise<void> {
  const cwd = process.cwd();
  const metaDir = await getMetaDir(cwd);

  if (!metaDir) {
    throw new Error('Not in a gogo-meta repository. Run "gogo init" first.');
  }

  const { config } = await readMetaConfig(cwd);
  const desired = Object.entries(config.projects);

  if (desired.length === 0) {
    output.success('Working copy already matches configuration');
    return;
  }

  const { urlToPath, ambiguousUrls } = await mapUrlsToCurrentPaths(metaDir, config.ignore);

  const moves: Move[] = [];
  const missing: string[] = [];
  const conflicts: Conflict[] = [];
  const ambiguous: string[] = [];

  for (const [projectPath, url] of desired) {
    const targetDir = join(metaDir, projectPath);

    if (await fileExists(targetDir)) {
      const targetRemote = await getRemoteUrl(targetDir);
      if (targetRemote !== url) {
        conflicts.push({ path: projectPath, found: targetRemote });
      }
      continue;
    }

    const currentPath = urlToPath.get(url);
    if (ambiguousUrls.has(url)) {
      ambiguous.push(projectPath);
    } else if (currentPath && currentPath !== projectPath) {
      moves.push({ from: currentPath, to: projectPath });
    } else {
      missing.push(projectPath);
    }
  }

  if (conflicts.length > 0) {
    for (const conflict of conflicts) {
      output.projectStatus(
        conflict.path,
        'error',
        `occupied by a different repository (found ${conflict.found ?? 'no remote'})`
      );
    }
    throw new Error(
      'Migration aborted: one or more target paths are occupied by a different repository'
    );
  }

  if (moves.length === 0 && missing.length === 0 && ambiguous.length === 0) {
    output.success('Working copy already matches configuration');
    return;
  }

  for (const move of moves) {
    if (options.dryRun) {
      output.info(`Would move ${move.from} → ${move.to}`);
      continue;
    }
    const targetDir = join(metaDir, move.to);
    await mkdir(dirname(targetDir), { recursive: true });
    await rename(join(metaDir, move.from), targetDir);
    await pruneEmptyParents(metaDir, move.from);
    await removeFromGitignore(metaDir, move.from);
    await addToGitignore(metaDir, move.to);
    output.projectStatus(move.to, 'success', `moved from ${move.from}`);
  }

  for (const projectPath of ambiguous) {
    output.warning(
      `${projectPath}: multiple working-copy directories share its repository URL — resolve manually`
    );
  }

  for (const projectPath of missing) {
    output.warning(`${projectPath} not found in working copy — run 'gogo git update' to clone`);
  }

  if (options.dryRun) {
    output.info(`Dry run: ${moves.length} move(s) pending`);
    return;
  }

  if (missing.length > 0 || ambiguous.length > 0) {
    process.exitCode = 1;
  }
}

export function registerMigrateCommand(program: Command): void {
  program
    .command('migrate')
    .description('Move/rename working-copy directories to match the configuration')
    .option('--dry-run', 'Show what would be moved without changing anything')
    .action(async (options: { dryRun?: boolean }) => {
      await migrateCommand({ dryRun: options.dryRun ?? false });
    });
}

File: internal/cli/migrate.go
package cli

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/daFish/gogo-meta/internal/config"
	"github.com/daFish/gogo-meta/internal/discover"
	"github.com/daFish/gogo-meta/internal/executor"
	"github.com/daFish/gogo-meta/internal/output"
	"github.com/spf13/cobra"
)

const (
	notAGogoRepoMsg     = `Not in a gogo-meta repository. Run "gogo init" first.`
	migrationAbortedMsg = "Migration aborted: one or more target paths are occupied by a different repository"
)

type migrateMove struct{ from, to string }
type migrateConflict struct {
	path  string
	found string // "" means no remote
}

func newMigrateCmd() *cobra.Command {
	var dryRun bool
	cmd := &cobra.Command{
		Use:   "migrate",
		Short: "Move/rename working-copy directories to match the configuration",
		RunE: func(_ *cobra.Command, _ []string) error {
			cwd, err := os.Getwd()
			if err != nil {
				return err
			}
			code, err := runMigrate(executor.NewShellExecutor(), cwd, dryRun)
			if err != nil {
				return err
			}
			if code != 0 {
				os.Exit(code)
			}
			return nil
		},
	}
	cmd.Flags().BoolVar(&dryRun, "dry-run", false, "Show what would be moved without changing anything")
	return cmd
}

func getRemoteURL(ex executor.Executor, dir string) string {
	res, err := ex.Execute(context.Background(), "git remote get-url origin", executor.Options{Cwd: dir})
	if err != nil || res.ExitCode != 0 {
		return ""
	}
	return strings.TrimSpace(res.Stdout)
}

func mapURLsToCurrentPaths(ex executor.Executor, metaDir string, ignore []string) (urlToPath map[string]string, ambiguous map[string]bool, err error) {
	var repoPaths []string
	repoPaths, err = discover.FindGitRepos(metaDir, ignore)
	if err != nil {
		return nil, nil, err
	}
	urlToPath = map[string]string{}
	ambiguous = map[string]bool{}
	for _, rp := range repoPaths {
		url := getRemoteURL(ex, filepath.Join(metaDir, rp))
		if url == "" {
			continue
		}
		if _, ok := urlToPath[url]; ok {
			ambiguous[url] = true
			continue
		}
		urlToPath[url] = rp
	}
	return urlToPath, ambiguous, nil
}

func pruneEmptyParents(metaDir, movedFrom string) {
	root, err := filepath.Abs(metaDir)
	if err != nil {
		return
	}
	dir, err := filepath.Abs(filepath.Dir(filepath.Join(metaDir, movedFrom)))
	if err != nil {
		return
	}
	for dir != root && strings.HasPrefix(dir, root+string(filepath.Separator)) {
		entries, err := os.ReadDir(dir)
		if err != nil {
			return
		}
		if len(entries) > 0 {
			return
		}
		if err := os.Remove(dir); err != nil {
			return
		}
		dir = filepath.Dir(dir)
	}
}

func runMigrate(ex executor.Executor, cwd string, dryRun bool) (int, error) {
	metaDir, err := config.GetMetaDir(cwd)
	if err != nil {
		return 0, err
	}
	if metaDir == "" {
		return 0, errors.New(notAGogoRepoMsg) //nolint:staticcheck // capitalized for JS parity
	}

	result, err := config.ReadMetaConfig(cwd, nil)
	if err != nil {
		return 0, err
	}
	cfg := result.Config

	if len(cfg.Projects) == 0 {
		output.Success("Working copy already matches configuration")
		return 0, nil
	}

	urlToPath, ambiguousURLs, err := mapURLsToCurrentPaths(ex, metaDir, cfg.Ignore)
	if err != nil {
		return 0, err
	}

	desired := make([]string, 0, len(cfg.Projects))
	for p := range cfg.Projects {
		desired = append(desired, p)
	}
	sort.Strings(desired)

	var moves []migrateMove
	var missing, ambiguous []string
	var conflicts []migrateConflict

	for _, projectPath := range desired {
		url := cfg.Projects[projectPath]
		targetDir := filepath.Join(metaDir, projectPath)

		if config.FileExists(targetDir) {
			targetRemote := getRemoteURL(ex, targetDir)
			if targetRemote != url {
				conflicts = append(conflicts, migrateConflict{path: projectPath, found: targetRemote})
			}
			continue
		}

		switch {
		case ambiguousURLs[url]:
			ambiguous = append(ambiguous, projectPath)
		case urlToPath[url] != "" && urlToPath[url] != projectPath:
			moves = append(moves, migrateMove{from: urlToPath[url], to: projectPath})
		default:
			missing = append(missing, projectPath)
		}
	}

	if len(conflicts) > 0 {
		for _, c := range conflicts {
			found := c.found
			if found == "" {
				found = "no remote"
			}
			output.ProjectStatus(c.path, "error", fmt.Sprintf("occupied by a different repository (found %s)", found))
		}
		return 0, errors.New(migrationAbortedMsg) //nolint:staticcheck // capitalized for JS parity
	}

	if len(moves) == 0 && len(missing) == 0 && len(ambiguous) == 0 {
		output.Success("Working copy already matches configuration")
		return 0, nil
	}

	for _, m := range moves {
		if dryRun {
			output.Info(fmt.Sprintf("Would move %s → %s", m.from, m.to))
			continue
		}
		targetDir := filepath.Join(metaDir, m.to)
		if err := os.MkdirAll(filepath.Dir(targetDir), 0o755); err != nil {
			return 0, err
		}
		if err := os.Rename(filepath.Join(metaDir, m.from), targetDir); err != nil {
			return 0, err
		}
		pruneEmptyParents(metaDir, m.from)
		if _, err := config.RemoveFromGitignore(metaDir, m.from); err != nil {
			return 0, err
		}
		if _, err := config.AddToGitignore(metaDir, m.to); err != nil {
			return 0, err
		}
		output.ProjectStatus(m.to, "success", fmt.Sprintf("moved from %s", m.from))
	}

	for _, p := range ambiguous {
		output.Warning(fmt.Sprintf("%s: multiple working-copy directories share its repository URL — resolve manually", p))
	}
	for _, p := range missing {
		output.Warning(fmt.Sprintf("%s not found in working copy — run 'gogo git update' to clone", p))
	}

	if dryRun {
		output.Info(fmt.Sprintf("Dry run: %d move(s) pending", len(moves)))
		return 0, nil
	}

	if len(missing) > 0 || len(ambiguous) > 0 {
		return 1, nil
	}
	return 0, nil
}
